src/domain_bridge/wait_for_qos_handler.hpp

```diff
@@ -0,0 +1,276 @@
+    auto opt_qos = get_topic_qos(topic, node);
+    if (opt_qos) {
+      const QosMatchInfo & qos = opt_qos.value();
+      callback(qos);
+      return;
+    }
+
+    {
+      std::lock_guard<std::mutex> lock(waiting_map_mutex_);
+      waiting_map_[node].push_back({topic, callback});
+
+      // If we already have a thread for this node, then notify that there is a new callback
+      auto existing_thread = waiting_threads_.find(node);
+      if (existing_thread != waiting_threads_.end()) {
+        existing_thread->second.second->notify_all();
+        return;
+      }
+    }
+
+    // If we made it this far, there doesn't exist a thread for waiting so we'll create one
+    // First, create a condition variable to notify the waiting thread when to wake up
+    auto cv = std::make_shared<std::condition_variable>();
+
+    // Create a thread that waits for a publisher to become available
+    auto invoke_callback_when_qos_ready = [this, node, cv]()
+      {
+        auto event = node->get_graph_event();
+        while (!this->shutting_down_.load()) {
+          // Wait for graph event
+          // Note, in case new publishers don't trigger a graph event we add a
+          // timeout so that we can still poll periodically for new publishers
+          node->wait_for_graph_change(event, std::chrono::seconds(1));
+          event->check_and_clear();
+          // Check if QoS is ready for any of the topics
+          {
+            std::unique_lock<std::mutex> lock(waiting_map_mutex_);
+            auto & topic_callback_vec = waiting_map_.at(node);
+            for (auto it = topic_callback_vec.begin(); it != topic_callback_vec.end(); ++it) {
+              const std::string & topic = it->first;
+              const auto callback = it->second;
+              auto opt_qos = this->get_topic_qos(topic, node);
+              if (opt_qos) {
+                const QosMatchInfo & qos = opt_qos.value();
+                callback(qos);
+                topic_callback_vec.erase(it--);
+              }
+            }
+            // If there are no more publishers to wait on,
+            // remove the entry from the map and terminate the thread
+            if (topic_callback_vec.size() == 0u) {
+              waiting_map_.erase(waiting_map_.find(node));
+            }
+
+            // It's only worth continuing if we have callbacks to handle,
+            // or we're shutting down
+            cv->wait(
+              lock,
+              [this, &topic_callback_vec]
+              {return topic_callback_vec.size() > 0u || this->shutting_down_.load();});
+          }
+        }
+      };
+    auto waiting_thread = std::make_shared<std::thread>(invoke_callback_when_qos_ready);
+    waiting_threads_[node] = {waiting_thread, cv};
+  }
+
+private:
+  using ThreadMap = std::unordered_map<
+    std::shared_ptr<rclcpp::Node>,
+    std::pair<std::shared_ptr<std::thread>, std::shared_ptr<std::condition_variable>>>;
```

Comment:
I don't mind, but it doesn't seem that you really need to wrap the `std::thread` and the `std::condition_variable` in a `std::shared_ptr` here

Reply:
`std::thread` and `std::condition_variable` are not copyable, so I think this is the only way I can store a list of them. Correct me if I'm wrong.

Reply:
> std::thread and std::condition_variable are not copyable, so I think this is the only way I can store a list of them. Correct me if I'm wrong.

you have to use `std::unordered_map::emplace()` to avoid that issue, e.g.:

```cpp
waiting_threads_.emplace(
  std::piecewise_construct,
  std::forward_as_tuple(node),
  std::forward_as_tuple(
    std::piecewise_construct, std::forward_as_tuple(invoke_callback_when_qos_ready), std::forward_as_tuple()));
```

so beautiful :joy:

Reply:
Though it's perhaps slightly less performant, using shared_ptr seems easier to understand :sweat_smile: